### ISSUE TITLE
Cannot access `content` property on result of octokit.repos.getContent()

### ISSUE THREAD

**Issue by wolfy1339** [octokit/rest.js#1971]
I ran into the same issue as above, though all of my `@octokit/types` references **are** to 6.x versions. 
Going from `@octokit/rest@18.0.9` to `@octokit/rest@18.0.10` broke a lot of TypeScript code.

For example, none of these types are available anymore. 
```ts
import type {
  SearchCodeResponseData,
  ReposGetResponseData,
  ReposGetBranchProtectionResponseData,
  ReposGetStatusChecksProtectionResponseData,
} from '@octokit/types';
```

Additionally, the Typescript types for `octokit.repos.getContent` seem to be broken. 
Previously it would return a `Promise<OctokitResponse<ReposGetContentResponseData>>` which had a data property with a content property.  

Now i get TypeScript errors trying to get to **content** - either through .data or directly. 

```
 'content' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string; sha: string; url: string; git_url: string; html_url: string; download_url: string; _links: { git: string; html: string; self: string; }; }[] | { ...; } | { ...; } | { ...; }'.
      Property 'content' does not exist on type '{ type: string; size: number; name: string; path: string; content?: string; sha: string; url: string; git_url: string; html_url: string; download_url: string; _links: { git: string; html: string; self: string; }; }[]'.
```

This seems like a breaking change that should require a major version bump

As an aside, i vastly prefer the named types of `@octokit/types@5` to the unnamed ones.

_Originally posted by @AndrewCsontos in https://github.com/octokit/rest.js/issues/1951#issuecomment-744449274_

**Comment by wolfy1339** [comment]
The issue seems to come from the fact that there is multiple types appended together using the `|` (OR) operator.

It all boils down to the [`Operation['response']` type](https://github.com/octokit/types.ts/blob/v6.1.1/src/generated/Endpoints.ts#L53)

**Comment by wolfy1339** [comment]
So, the bug stems from the [OpenAPI types](https://github.com/octokit/openapi-types.ts/blob/main/generated/types.ts#L12349) themselves.

Due to how the types are constructed, you cannot use the `content` property on the response, because `components["schemas"]["content-submodule"]` does not include that property.

**Comment by gr2m** [comment]
I don't think this is a bug. The [Get repository content](https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#get-repository-content) API endpoint can have different responses depending on exists at the given path, hence the union. If the `response.data.content` property was typed as if it always existed before, it was wrong. You have to check if it exists before using it

**Comment by wolfy1339** [comment]
@AndrewCsontos this should fix your issue
```ts
import { Octokit } from '@octokit/rest';
const octokit = new Octokit();

async function foo() {
  const content = await octokit.repos.getContent();
  if (typeof content.data !== undefined) {
    content.data
  }
}
```

**Comment by wolfy1339** [comment]
> If the response.data.content property was typed as if it always existed before, it was wrong. You have to check if it exists before using it

Because the `content` property does not exist on all the type definitions, TypeScript doesn't let us use it, hence the error, even if you try to check if it exists.
Example:
```ts
import { Octokit } from '@octokit/rest';
const octokit = new Octokit();

async function foo() {
  const content = await octokit.repos.getContent();
  if (typeof content.content === 'string') {
    console.log(content.content)
  }

  if (typeof content.content !== undefined) {
    console.log(content.content)
  }

}
```
Both of those checks fail with the error reported by  @AndrewCsontos, that is the result of the union.

Unfortunately, in TypeScript, all types in the union must contain the properties or else the code will error (ref: https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#unions-with-common-fields)

**Comment by gr2m** [comment]
Okay I see. `{ data } = await response.data.content()` types `data` as unknown. 

This sounds like a problem with https://github.com/drwpow/openapi-typescript. Could you like to file an issue there? They use GitHub's OpenAPI spec as one of their test cases

**Comment by wolfy1339** [comment]
> Okay I see. `{ data } = await response.data.content()` types data as unknown.
That's only for the 403, and 404 status codes. The 302 status code also has an invalid type (according to the openapi spec)

If we stay at the 200 status code, the types are okay. But the problem is the same, since they don't all have the `content` property that property will always error

**Comment by gr2m** [comment]
> If we stay at the 200 status code, the types are okay

Hmm I thought I filtered out all non 2xx response types in https://github.com/octokit/types.ts:
https://github.com/octokit/types.ts/blob/ecdc06f09db8b9785e06f8b07cb6f14f4d57e8b9/src/generated/Endpoints.ts#L67-L102

Or am I missing something?

**Comment by wolfy1339** [comment]
In `@octokit/openapi-types`, `paths['/repos/{owner}/{repo}/contents/{path}']['get']['responses']`, you can see there what I'm talking about with the non-2xx responses being typed as `unknown` or `never`.

Now, the 2xx responses are typed accordingly, and they aren't typed as `unknown` or `never`

**Comment by wolfy1339** [comment]
I've been looking around the types and there's 2 problems here:
1. `response.data` can be an `Array` when the path is a directory
2. The other types for `response.data` don't include the `content` property (when path is a git submodule or a symlink)

**Comment by wolfy1339** [comment]
@AndrewCsontos a workaround for now, is to add `as components["schemas"]["content-file"]` where `components` is imported from `@octokit/openapi-types`, whenever you want to access the `content` property.

Ex:
```ts
import { Octokit } from '@octokit/rest';
import { paths, components } from '@octokit/openapi-types'
const octokit = new Octokit();

type GetRepoContentResponseDataFile = components["schemas"]["content-file"]
async function foo() {
  const { data } = await octokit.repos.getContent();
  if (!Array.isArray(data)) {
    const foo = data as GetRepoContentsDataFile

    if (typeof foo.content !== undefined) {
      console.log(foo.content)
    }
  }
}
```

**Comment by AndrewCsontos** [comment]
Thanks for the tip. I'll probably stay on 18.0.9 for a while. 
I also still think the type switch-over is a breaking change that should result in bumping octokit/rest to 19.x

**Comment by wolfy1339** [comment]
@dominguezcelada Do you have any ideas how to fix this issue?

**Comment by dominguezcelada** [comment]
> @dominguezcelada Do you have any ideas how to fix this issue?

Tomorrow I will take a look into it and see if I can help ;)

**Comment by dominguezcelada** [comment]
> > @dominguezcelada Do you have any ideas how to fix this issue?
> 
> Tomorrow I will take a look into it and see if I can help ;)

I don't find an easy solution for the problem. Searching on the Internet I found some implementations of `OneOf` or `ExclusiveUnion` like this but looks super complex to me:
* [Source](https://dev.to/maxime1992/implement-a-generic-oneof-type-with-typescript-22em)
* [Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAaghgGwK4QPIDMA8qBGArAPigF4pc8BtAawhAHt0z8BdAbgFgAoUSMgOzR8EIbPgA0UANJQIAD2AQ+AEwDOUGvUbkipAN5RqtKAEs+UAKKyAxsiURMGhkzwTJBZgC4ofJAgRQAXygAMih9QxATM0lPZwoYwI5ucGhUAQwAIRBJWlFCEjCDDSj1WidyWLTBYTyJDSIApLkwOgAnYCgeVPSsABUdWEQUDGwerJyRfoIkrlMFVvQ4K2gAEToAczCuKB24DxVgVtN1rgDZvnnF5agAdToERl1tnZx9w+PT88ul6HM4dYQ0CenB2UCsbyOfBOnDOnDmEAWPygAFk6EgVEDnlAlBCPjCuFwulAAIJ8YwAW0QBWBoOMOKgB0h6ySoL4cHJEC8jOOSSCoSqIxpOyUGy8a2ZWIA7vd0F47g8WTsIP9AV4-gCIIqoOS0RivKj0ZrTtMCXCLgirtAAAoIOAgBFbEE7Olc95QrVkqxUNkc11MrWAgBuEAQXh85JwCK1cDJlNDJNjiF5pqsdD4BygYFt9taXhtdodeixLqgAHIswXWgBaAAMpbEWM93vZnLLlNkFIgAEYAJw9gBM9axQZDXhrDadUBjFMQXiFoJM9NL07jtaHk9ZLa8pYA4gALOgHdcL4WisJ7MtwUsBCcnmQq1u6V5lnDX28nnWGsO+BDvhfSh5vz8P9YVhLggA)

What about wrapping common properties in a `type` and extending it with `type { [key:string]: any}` for the moment to unblock users and we ping TypeScript team? With more time we can research on existing issues for this OneOf implementation and what's the best approach here.

**Comment by wolfy1339** [comment]
It seems we aren't the only ones who have come across this problem. There seems to be a  workaround suggested by the TS devs that could work
https://github.com/microsoft/TypeScript/issues/23928

**Comment by dominguezcelada** [comment]
> It seems we aren't the only ones who have come across this problem. There seems to be a workaround suggested by the TS devs that could work
> [microsoft/TypeScript#23928](https://github.com/microsoft/TypeScript/issues/23928)

But this solution still relies on users (using `in` operator) instead of solving the problem at type creation level right?

Surfing a bit on open issues I find this but still not clear the status and the approach to follow:
* https://github.com/microsoft/TypeScript/issues/12936#issuecomment-284590083

Should we move forward making properties `optional` so we remove friction here and create a related issue where we can spike a proper solution for this?

**Comment by wolfy1339** [comment]
That won't really help much. The types are only common properties, so we'd have to add that property to all the different openapi types.

This just seems like a problem that can't really be fixed.

**Comment by dominguezcelada** [comment]
> That won't really help much. The types are only common properties, so we'd have to add that property to all the different openapi types.
> 
> This just seems like a problem that can't really be fixed.

Let me research a little bit more and open a StackOverflow post explaining the issue. But for sure there are multiple issues opened in TypeScript repo for this.

**Comment by wolfy1339** [comment]
> Let me research a little bit more and open a StackOverflow post explaining the issue. But for sure there are multiple issues opened in TypeScript repo for this.

Did you get any answers on your StackOverflow post?

**Comment by oscard0m** [comment]
> > Let me research a little bit more and open a StackOverflow post explaining the issue. But for sure there are multiple issues opened in TypeScript repo for this.
> 
> Did you get any answers on your StackOverflow post?

Hey @wolfy1339 sorry for the delay here. After checking a bit deeper the issue and reading some StackOverflow / Github issues open and googling a bit, I think the following solution would be the right one, let me know what do you think: 

What [TypeScript handbook says](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#discriminating-unions ): **we need a to have a single field which uses literal types which you can use to let TypeScript narrow down the possible current type**

Checking the different types for `content` I think the `type` field is the one we should be using:

https://github.com/octokit/openapi-types.ts/blob/main/generated/types.ts#L25896-L25962 

so, instead of:

```typescript
"content-directory": {
	type: string;
    ...
}[];

"content-file": {
	type: string;
    ...
};

"content-symlink": {
	type: string;
	...
};

"content-submodule": {
	type: string;
	...
};
```
we should use:

```typescript
"content-directory": {
      type: "dir";
      ...
}[];

"content-file": {
    type: "file";
	...
};

"content-symlink": {
	type: "symlink";
	...
};

"content-submodule": {
	type: "submodule";
	...
};
```

Which is already like this checking the examples in GitHub docs https://docs.github.com/en/free-pro-team@latest/rest/reference/repos#get-repository-content but checking their OpenAPI it is not like this: https://raw.githubusercontent.com/github/rest-api-description/main/descriptions/api.github.com/api.github.com.json

If you think this is the issue... Does this mean we (I can do it myself if you want) need to open an issue in Github's OpenAPI (I'm still not familiar with the flow and pieces)? @gr2m @wolfy1339 

P.S.: Tagging @G-Rath just in case you are interested in following up or adding your thoughts on this.

**Comment by wolfy1339** [comment]
> If you think this is the issue... Does this mean we (I can do it myself if you want) need to open an issue in Github's OpenAPI (I'm still not familiar with the flow and pieces)? @gr2m @wolfy1339

Yes, we should open an issue there, and reference back to this one.

**Comment by G-Rath** [comment]
@oscard0m that sounds about what I'd expect - it's the same reason I restructured our schemas to be per-action.

As you've found in the handbook, you need to have a discriminating property to be able to properly type narrow, which your proposal adds :)

**Comment by wolfy1339** [comment]
Blocked by: https://github.com/github/rest-api-description/issues/165

**Comment by gr2m** [comment]
Great research  @oscard0m!

> ```ts
> "content-directory": {
>       type: "dir";
>       ...
> }[];
> ```

 Note that in case of a directory response, the items can be one of the four possible types: `file`, `dir`, `symlink`, `submodule`, not just `dir`

**Comment by rudy750** [comment]
This is still happening to me. Here is what I did:
```
async function foo() {
  const response = await octokit.repos.getContent(
    {
          owner: 'me',
          repo: 'mine',
          path: 'foo/bar.json',
     })
);
    const { content } = { ...response.data };
}
```

then I have my content in base64. 
Hope that helps
